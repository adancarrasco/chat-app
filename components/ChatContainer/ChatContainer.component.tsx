import React, { useEffect, useReducer } from "react";
import { View } from "react-native";
import database from "@react-native-firebase/database";

import * as mockData from "../../mockDB.json";
import Message from "../Message";
import ActionButton from "../ActionButton";
import chatReducer, {
  ADD_TO_CONVERSATION,
  IThread,
  FILL_THREAD,
  REMOVE_FROM_CONVERSATION
} from "../../reducers/chat";
import ImageMessage from "../ImageMessage";

const initialState = { previous: [], conversation: [], thread: {} };

export default function ChatContainer() {
  const uid = "4szFlgouIMJ75tQi3xey";
  const [state, dispatch] = useReducer(chatReducer, initialState);
  const TAP_TO_CONTINUE = "Tap to continue";

  useEffect(() => {
    // Create reference
    const ref = database().ref(`/stories/${uid}`);
    let thread: IThread;

    function initFlow() {
      dispatch({
        type: ADD_TO_CONVERSATION,
        payload: {
          ...thread.storyThread[thread.firstMessage],
          key: thread.firstMessage,
          shouldAnimate: true
        }
      });
      dispatch({ type: FILL_THREAD, payload: thread.storyThread });
    }

    ref
      .once("value", snapshot => {
        thread = snapshot.val();
        initFlow();
      })
      .catch(error => console.warn(error));

    setTimeout(() => {
      if (!thread) {
        thread = mockData.stories[uid];
        initFlow();
      }
    }, 2000);
  }, [uid]);

  return (
    <View>
      <ActionButton
        onTouchEnd={() => {
          dispatch({
            type: REMOVE_FROM_CONVERSATION
          });
        }}
      />
      <View style={{ height: "100%" }}>
        {state.conversation.length > 0 &&
          state.conversation.map(item => {
            const { isLocalUser, name, message, key, shouldAnimate } = item;
            return item.type === "text" ? (
              <Message
                isLocalUser={isLocalUser}
                userName={name}
                message={message}
                key={key}
                shouldAnimate={shouldAnimate}
              />
            ) : (
              <ImageMessage
                shouldAnimate={shouldAnimate}
                isLocalUser={isLocalUser}
                key={key}
              />
            );
          })}
      </View>
      <ActionButton
        isBottomButton={true}
        onTouchEnd={() => {
          const count = state.conversation.length - 1;
          if (!state.conversation[count].next) return;
          dispatch({
            type: ADD_TO_CONVERSATION,
            payload: {
              ...state.thread[state.conversation[count].next[0]],
              key: state.conversation[count].next[0],
              shouldAnimate: true
            }
          });
        }}
      >
        {TAP_TO_CONTINUE}
      </ActionButton>
    </View>
  );
}
