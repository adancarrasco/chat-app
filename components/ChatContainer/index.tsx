export { default } from "./ChatContainer.component";
