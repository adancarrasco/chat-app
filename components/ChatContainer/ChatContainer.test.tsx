import React from "react";
import renderer from "react-test-renderer";
import ChatContainer from "./ChatContainer.component";

describe("<ChatContainer /> rendering", () => {
  it("Matches snapshot", () => {
    const tree = renderer.create(<ChatContainer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
