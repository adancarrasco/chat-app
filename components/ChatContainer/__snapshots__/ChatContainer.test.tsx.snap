// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatContainer /> rendering Matches snapshot 1`] = `
<View>
  <View
    onTouchEnd={[Function]}
    style={
      Object {
        "alignItems": "flex-end",
        "borderRadius": 8,
        "flexDirection": "row",
        "height": "50%",
        "justifyContent": "center",
        "opacity": 0,
        "padding": 10,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 10,
      }
    }
  >
    <Text />
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  />
  <View
    onTouchEnd={[Function]}
    style={
      Object {
        "alignItems": "flex-end",
        "borderRadius": 8,
        "flexDirection": "row",
        "height": "50%",
        "justifyContent": "center",
        "opacity": 0,
        "padding": 10,
        "position": "absolute",
        "top": "50%",
        "width": "100%",
        "zIndex": 10,
      }
    }
  >
    <Text>
      Tap to continue
    </Text>
  </View>
</View>
`;
